src/CoreWCF.Primitives/src/CoreWCF/Security/SecuritySessionServerSettings.cs

```diff
@@ -1956,6 +1968,568 @@ public Task OpenAsync(CancellationToken token)
+               // bool didOutputSessionClose = WaitForOutputSessionClose(ServiceDefaults.CloseTimeout, out wasAborted);
+                if (wasAborted)
+                {
+                    return;
+                }
+
+                if (!didOutputSessionClose)
+                {
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperWarning(new TimeoutException(SR.Format(SR.ServiceSecurityCloseOutputSessionTimeout, ServiceDefaults.CloseTimeout)));
+                }
+
+                CloseCore(token);
+                Settings.RemoveSessionChannel(_session.Id);
+            }
+
+            private void DetermineCloseOutputSessionMessage(out bool sendClose, out bool sendCloseResponse, out Message pendingCloseResponseMessage, out RequestContext pendingCloseRequestContext)
+            {
+                sendClose = false;
+                sendCloseResponse = false;
+                pendingCloseResponseMessage = null;
+                pendingCloseRequestContext = null;
+                lock (LocalLock)
+                {
+                    if (!_isOutputClosed)
+                    {
+                        _isOutputClosed = true;
+                        if (_receivedClose)
+                        {
+                            if (_closeResponseMessage != null)
+                            {
+                                pendingCloseResponseMessage = _closeResponseMessage;
+                                pendingCloseRequestContext = _closeRequestContext;
+                                _closeResponseMessage = null;
+                                _closeRequestContext = null;
+                                sendCloseResponse = true;
+                            }
+                        }
+                        else
+                        {
+                            sendClose = true;
+                            _sentClose = true;
+                        }
+                        _outputSessionCloseHandle.Reset();
+                    }
+                }
+            }
+
+            private Task CloseOutputSessionAsync(CancellationToken token)
+            {
+                bool sendClose = false;
+                bool sendCloseResponse = false;
+                Message pendingCloseResponseMessage;
+                RequestContext pendingCloseRequestContext;
+                try
+                {
+                    DetermineCloseOutputSessionMessage(out sendClose, out sendCloseResponse, out pendingCloseResponseMessage, out pendingCloseRequestContext);
+                    if (sendCloseResponse)
+                    {
+                        bool cleanupCloseState = true;
+                        try
+                        {
+                            SendCloseResponse(pendingCloseRequestContext, pendingCloseResponseMessage, token);
+                            cleanupCloseState = false;
```

Comment:
This method is still synchronous. The SendCloseResponse method should be async. Looking at that method it looks like there's an existing bug that needs to be fixed. The SendCloseResponse method calls SendMessage which should be async but isn't. SendMessage calls                     `requestContext.ReplyAsync(message, token)` followed by `requestContext.CloseAsync(token);` without awaiting between them. The request context could get closed before the reply has been sent and this could result in difficult to track down exceptions. So can you fix SendMessage to be async and await on it's calls correctly and then fix all the methods up the chain.

Reply:
Fixed SendClose as async.. but kept SendRenewFault sync with discard for SendMessage call... the reason for that this is in the  receive request path after which rest things get into picture.. let me know if there is better way.. open to suggestion :)

Reply:
Can you add an inline comment in SendRenewFault acknowledging it's sync over async and giving the rationale of why that's okay.